index.d.ts

```diff
@@ -0,0 +1,70 @@
+import { ComponentClass, Component, StatelessComponent, ReactNode } from 'react';
+import { Store, Dispatch, ActionCreator } from 'redux';
+
+interface ComponentDecorator<TOriginalProps, TOwnProps> {
+  (component: ComponentClass<TOriginalProps>|StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+}
+
+/**
+ * Following 3 functions cover all possible ways connect could be invoked
+ *
+ * - State: Redux state interface (the same one used by Store<S>)
+ * - TStateProps: Result of MapStateToProps
+ * - TDispatchProps: Result of MapDispatchToProps
+ * - TOwnProps: Props passed to the wrapping component
+ */
+export function connect<State, TOwnProps>(): ComponentDecorator<{ dispatch: Dispatch<State> }, TOwnProps>;
+
+export function connect<State, TOwnProps, TStateProps>(
+  mapStateToProps: MapStateToProps<State, TStateProps, TOwnProps>,
+): ComponentDecorator<TStateProps & { dispatch: Dispatch<State> }, TOwnProps>;
+
+export function connect<State, TOwnProps, TStateProps, TDispatchProps>(
+  mapStateToProps: MapStateToProps<State, TStateProps, TOwnProps>|null,
+  mapDispatchToProps: MapDispatchToPropsFunction<State, TDispatchProps, TOwnProps>|TDispatchProps
```

Comment:
This is the road taken in DT: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/react-redux/react-redux.d.ts#L73

I was against it because you lose the most important part (for me at least): make sure your `TDispatchProps` matches what you are returning from `mapDispatchToProps`.

With DT's version, this doesn't error:
```typescript
import { checkout } from 'actions'

interface TDispatchProps {
  checkout: (cartId: string) => void // should be number
  // or
  checout: (cartId: number) => void // typo
}
```

This works and check for both issues (TDispatchProps and hash of ActionCreator): https://github.com/reactjs/react-redux/pull/538/commits/670c85146170ff712587782303a2f7afbe5fbda3

Done